In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
sns.set()

In [2]:
raw_data = pd.read_csv('2.02. Binary predictors.csv')
raw_data.head()

SAT Admitted  Gender
0  1363       No    Male
1  1792      Yes  Female
2  1954      Yes  Female
3  1653       No    Male
4  1593       No    Male

In [3]:
data = raw_data.copy()
data['Admitted'] = data['Admitted'].map({'Yes':1,'No':0})
data['Gender'] = data['Gender'].map({'Female':1,'Male':0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [4]:
y = data['Admitted']
x1 = data[['SAT','Gender']]

In [5]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x1,y,test_size=0.1)

In [8]:
reg = LogisticRegression()
reg.fit(x_train,y_train)

LogisticRegression()

In [9]:
reg.intercept_

array([-67.81597546])

In [26]:
reg.coef_

array([[0.04065225, 1.11866566]])

## Confusion matrix on sklearn

In [27]:
y_predict = reg.predict(x_train)
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_train,y_predict)
cm_df = pd.DataFrame(cm)
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df.index = ['Actual 0','Actual 1']
cm_df

Predicted 0  Predicted 1
Actual 0           61            4
Actual 1            5           81

In [28]:
accuracy_train = (cm[0,0]+cm[1,1])/cm.sum()
accuracy_train

0.9403973509933775

In [29]:
cm

array([[61,  4],
       [ 5, 81]], dtype=int64)

In [30]:
f = 1
m = 0
result = (np.exp(reg.coef_[0,1])*(f-m)).round(2)
print('The odds of female to get admitted is {} times greater than the odds of males to get admitted in this dataset'.format(result))
    

The odds of female to get admitted is 3.06 times greater than the odds of males to get admitted in this dataset


### Test the model

In [32]:
y_predict_test = reg.predict(x_test)

In [33]:
cm_test = confusion_matrix(y_test,y_predict_test)
cm_df_test = pd.DataFrame(cm_test)
cm_df_test.columns = ['Predicted 0','Predicted 1']
cm_df_test.index = ['Actual 0','Actual 1']
cm_df_test

Predicted 0  Predicted 1
Actual 0            7            2
Actual 1            0            8

In [34]:
accuracy_train = (cm_test[0,0]+cm_test[1,1])/cm_test.sum()
accuracy_train

0.8823529411764706

In [36]:
reg.score(x_test,y_test)

0.8823529411764706

### How to Predict

In [39]:
SAT= int(input())
Gender = int(input())
yhat = reg.intercept_+reg.coef_[0,0]*SAT+reg.coef_[0,1]*Gender
p = 1/(1+np.exp(-yhat))
if p>0.5:
    print('Probability:',(p*100))
    print('1')
else:
    print('Probability:',p*100)
    print('0')
    

1750
1
Probability: [98.83890048]
1


In [40]:
98.8-96.5

2.299999999999997

### For the same SAT number 1750,
#### Probability of female to get admitted is = 98.8
#### Probability of male to get admitted is = 96.5
#### 98.8-96.5 = 2.3 times greater for female (almost accurate)